examples/ReinterpretLog.hs

```diff
@@ -0,0 +1,159 @@
+{-# LANGUAGE DerivingStrategies         #-}
+{-# LANGUAGE DerivingVia                #-}
+{-# LANGUAGE FlexibleContexts           #-}
+{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE GADTs                      #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE InstanceSigs               #-}
+{-# LANGUAGE KindSignatures             #-}
+{-# LANGUAGE LambdaCase                 #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+{-# LANGUAGE RankNTypes                 #-}
+{-# LANGUAGE TypeOperators              #-}
+{-# LANGUAGE UndecidableInstances       #-}
+
+module ReinterpretLog where
+
+import Control.Effect
+import Control.Effect.Carrier
+import Control.Effect.Sum
+import Control.Monad.IO.Class     (MonadIO(..))
+import Control.Monad.Trans.Reader (ReaderT(..))
+import Data.Coerce                (coerce)
+import Data.Function              ((&))
+import Data.Kind                  (Type)
+import Prelude                    hiding (log)
+
+
+--------------------------------------------------------------------------------
+-- The application
+--------------------------------------------------------------------------------
+
+-- Our structured log message.
+data Message
+  = Debug String
+  | Info String
+
+-- Render a structured log message as a string.
+renderLogMessage :: Message -> String
+renderLogMessage = \case
+  Debug message -> "[debug] " ++ message
+  Info  message -> "[info] "  ++ message
+
+-- The application: it logs two messages, then quits.
+application :: (Carrier sig m, Member (Log Message) sig) => m ()
+application = do
+  log (Debug "debug message")
+  log (Info "info message")
+
+-- The application runner. Interpret the application by:
+--
+-- * Reinterpreting 'Log Message' effects as 'Log String' effects.
+-- * Interpreting 'Log String' effects by printing to stdout.
+runApplication :: IO ()
+runApplication =
+  application
+    & reinterpretLog renderLogMessage
+    & runLogStdout
+    & runM
+
+
+--------------------------------------------------------------------------------
+-- The logging effect
+--------------------------------------------------------------------------------
+
+-- Log an 'a', then continue with 'k'.
+data Log :: Type -> (Type -> Type) -> Type -> Type where
+  Log :: a -> k -> Log a m k
+  deriving stock (Functor)
+
+instance Effect (Log a) where
+  handle ::
+       Functor f
+    => f ()
+    -> (forall x. f (m x) -> n (f x))
+    -> Log a m (m b)
+    -> Log a n (n (f b))
+  handle state handler (Log x k) =
+    Log x (handler (k <$ state))
+
+instance HFunctor (Log a) where
+  hmap :: (forall x. m x -> n x) -> Log a m k -> Log a n k
+  hmap _ =
+    coerce
+
+-- Log an 'a'.
+log :: (Carrier sig m, Member (Log a) sig) => a -> m ()
+log x =
+  send (Log x (pure ()))
+
+
+--------------------------------------------------------------------------------
+-- The logging effect carriers
+--------------------------------------------------------------------------------
```

Comment:
In order to define a spec to show off how this operates, I’d suggest defining a carrier which simply collects the logged messages (of whatever type), and then testing the list of collected (transformed) messages against the expected strings.